File: acm-icpc/2013/world_final/D.cc
#include <iostream>
#include <cstdio>
#include <algorithm>
#include <vector>
#include <map>
using namespace std;
const unsigned long long maxn = (1LL<<63);
bool flag[1000] = {0};
vector<int> prime;

unsigned long long c[100][100];
void init() {
  int total = 1;
  for (int i = 2; i < 1000 ; i ++) {
    if (flag[i] == 0) {
      prime.push_back(i);
      for (int j = i * 2; j < 1000; j += i) flag[j] = 1;
    }
  }

  for (int i = 0; i < 100; i ++) {
    c[i][0] = 1;
    for (int j = 1; j <= i; j ++) {
      c[i][j] = 0;
      if (i > 0 && j > 0)
        c[i][j] += c[i-1][j-1];
      if (i > 0)
        c[i][j] += c[i-1][j];
    }
  }
}

map<unsigned long long, unsigned long long> cache;
int cnt[210];
const int max_n = 65;
void dfs(int deep, unsigned long long now, int max_, int n) {
  if (now <= 0)
    return;
  if (max_ == 0) {
    unsigned long long total = 1;
    for (int i = 0; i < deep; i ++) {
      if (total <= maxn / c[n][cnt[i]]) {
        total *= c[n][cnt[i]];
        n -= cnt[i];
      }
      else {
        return;
      }
    }
    if (cache[total] == 0 || cache[total] > now)
      cache[total] = now;
    return;
  }

  unsigned long long tmp = 1LL;
  for (int i = 0; (i <= max_) && (tmp  <= maxn / now) ; i ++) {
    cnt[deep] = i;
    dfs(deep + 1, now * tmp, i, n + i);
    if (tmp <= maxn / prime[deep])
      tmp *= prime[deep];
    else
      return;
  }
}
int main() {
  // cout << maxn << endl;
  init();
  // cout << prime.size() << endl;
  dfs(0,1,100,0);
  // cout << cache.size() << endl;
  unsigned long long m;
  cache[1] = 2;

  while (cin >> m) {
    cout << m << " " << cache[m] << endl;
  }
  return 0;
}

File: acm-icpc/2013/world_final/D.java
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;
public class D {
  public static HashMap<BigInteger, BigInteger> cache = new HashMap<BigInteger, BigInteger>();
  public static BigInteger[][] c = new BigInteger[100][100];
  public static boolean[] flags = new boolean[1000];
  public static ArrayList<BigInteger> primes = new ArrayList<BigInteger>();
  public static int count[] = new int[100];
  public static BigInteger biggest;
  public static void init() {
    primes.clear();
    for (int i = 0; i < flags.length; i ++) flags[i] = false;
    for (int i = 2; i < flags.length; i ++) {
      if (!flags[i]) {
        primes.add(new BigInteger(Integer.toString(i)));
        for (int j = i * i; j < flags.length; j += i) flags[j] = true;
      }
    }

    for (int i = 0; i < c.length; i ++) {
      c[i][0] = new BigInteger("1");
      for (int j = 1; j <= i; j ++) {
        c[i][j] = new BigInteger("0");
        if (i > 0 && j > 0 && c[i - 1][j - 1] != null)
          c[i][j] = c[i][j].add(c[i -1][j - 1]);
        if (i > 0 && c[i - 1][j] != null)
          c[i][j] = c[i][j].add(c[i - 1][j]);
      }
    }
  }
  public static void dfs(int deep, BigInteger now, int max, int n) {
    if (max == 0) {
      if (now.compareTo(biggest.ONE) < 1)
        return;
      BigInteger total = new BigInteger("1");
      for (int i = 0; i < deep; i ++) {
        total = total.multiply(c[n][count[i]]);
        if (total.compareTo(biggest) > 0)
          return;
        n -= count[i];
      }

      if (!cache.containsKey(total) || cache.get(total).compareTo(now) > 0)
        cache.put(total, now);
      return;
    }
    BigInteger num = new BigInteger("1");
    BigInteger tmp = null;
    for (int i = 0; i <= max; i ++) {
      tmp = now.multiply(num);
      if (tmp.compareTo(biggest) <= 0) {
        count[deep] = i;
        dfs(deep + 1, tmp, i, n + i);
      } else {
        break;
      }
      num = num.multiply(primes.get(deep));
    }
  }
  public static void main(String[] args) {
    init();
    biggest = new BigInteger(Long.toString(Long.MAX_VALUE));
    dfs(0, BigInteger.ONE, 66, 0);
    Scanner scanner = new Scanner(System.in);
    BigInteger k = null;
    while (scanner.hasNext()) {
      k = new BigInteger(scanner.next());
      System.out.print(k);
      System.out.print(" ");
      System.out.println(cache.get(k));
    }
  }
}
